numpy/core/src/umath/ufunc_object.c

```diff
@@ -2299,19 +2387,104 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
      * (Just checks core; broadcast dimensions are tested by the iterator.)
      */
     for (i = 0; i < nop; i++) {
-        if (op[i] != NULL && PyArray_NDIM(op[i]) < core_num_dims[i]) {
-            PyErr_Format(PyExc_ValueError,
+        /* TODO: store _n_required after parsing, ahead-of-time */
+        int n_required = ufunc->core_num_dims[i];
+        if (n_required > 0) {
+            /* can be less if dim is flexible */
+            for (j = ufunc->core_offsets[i];
+                     j < ufunc->core_offsets[i] + core_num_dims[i]; j++) {
+                int dim_index = ufunc->core_dim_ixs[j];
+                n_required -= ufunc->core_dim_flexible[dim_index];
+            }
+        }
+        if (op[i] != NULL) {
+            n_dims_used[i] = PyArray_NDIM(op[i]);
+            if (n_dims_used[i] < n_required) {
+                PyErr_Format(PyExc_ValueError,
                          "%s: %s operand %d does not have enough "
                          "dimensions (has %d, gufunc core with "
                          "signature %s requires %d)",
                          ufunc_name,
                          i < nin ? "Input" : "Output",
-                         i < nin ? i : i - nin,
-                         PyArray_NDIM(op[i]),
-                         ufunc->core_signature,
-                         core_num_dims[i]);
-            retval = -1;
-            goto fail;
+                         i < nin ? i : i - nin, PyArray_NDIM(op[i]),
+                         ufunc->core_signature, n_required);
+                retval = -1;
+                goto fail;
+            }
+            if (n_dims_used[i] < core_num_dims[i]) {
+                int flexible_used = core_num_dims[i] - n_dims_used[i];
+                /* one (or more?) flexible signature names have been used. Mark them */
+                for (j = ufunc->core_offsets[i];
+                     j < ufunc->core_offsets[i] + core_num_dims[i];
+                     j++) {
+                    int dim_index = ufunc->core_dim_ixs[j];
+                    if (flexible_activated[dim_index] < 0) {
+                        flexible_activated[dim_index] = 1;
+                    }
+                    flexible_used -= flexible_activated[dim_index];
```

Comment:
Is there an idea that `flexible_activated` can be anything else but 0 or 1 here? If not, I think this part is clearer as
```
if ((flexible_activated[dim_index] < 0) {
    flexible_activated[dim_index] = 1;
    flexible_used--;
    if (flexible_used == 0) {
        break;
    }
}
```

Reply:
`flexible_activated[dim_index]` may have been set to `1` by a previously encountered use of that `dim_index`, i.e. if the signature is (m?,k), (k, ?m) -> (m?, m?), when we reach the second `nin` we want to make sure the right-hand shape matches the first `nin`'s left-hand shape. So the decrement-and-check needs to occur outside the `if`